server/src/main/java/org/opensearch/search/fetch/FetchSearchResult.java

```diff
@@ -64,9 +72,24 @@ public FetchSearchResult(StreamInput in) throws IOException {
         hits = new SearchHits(in);
     }
 
+    public FetchSearchResult(InputStream in) throws IOException {
+        this.fetchSearchResultProto = FetchSearchResultProto.FetchSearchResult.parseFrom(in);
+        contextId = new ShardSearchContextId(
+            this.fetchSearchResultProto.getContextId().getSessionId(),
+            this.fetchSearchResultProto.getContextId().getId()
+        );
+        SearchHitsProtobufSerializer protobufSerializer = new SearchHitsProtobufSerializer();
+        hits = protobufSerializer.createSearchHits(new ByteArrayInputStream(this.fetchSearchResultProto.getHits().toByteArray()));
+    }
+
     public FetchSearchResult(ShardSearchContextId id, SearchShardTarget shardTarget) {
         this.contextId = id;
         setSearchShardTarget(shardTarget);
+        this.fetchSearchResultProto = FetchSearchResultProto.FetchSearchResult.newBuilder()
```

Comment:
Transport-specific implementation needs to live in its own namespaces/classes.

Reply:
This is not a transport specific implementation though. This is the response/request specific serialization and deserialization. Transport specific protobuf implementation will be separate which is sent over the wire of which this message is a part. Transport specific protobuf will be a part of the next PR.

Reply:
The way it's written the request/response can serialize its way two ways, however the serialization primitives are all over the class. I think there are a few options (and possibly more).

1. Multiple `serialize` methods which take different stream types, and no locals anywhere else in the class.
2. Different `serialize` methods in the style of `toString()`, so `serializeToNative(..)` or `serializeToProtobuf(..)`.
3. A base class `FetchResult` and child classes `NativeFetchResult` and `ProtobufFetchResult` that both have a `serialize` method that comes from `NativeSerializable` and `ProtobufSerializable`.
4. A helper serializer class `NativeFetchResultSerializer` and `ProtobufFetchResultSerializer`.

Maybe @reta has better ideas?

Reply:
Thanks @dblock definitively +1 (https://github.com/opensearch-project/OpenSearch/pull/13178/files#r1566328274)

Reply:
But `DocumentField` used to implement Writeable, we changed that. The approach with inheritance is an option but a) it won't scale b) it is inconsistent with others (`DocumentField`)

Reply:
The difference between `DocumentField` and `QuerySearchResult` (or similar) for example is that `DocumentField` is just a model class. While `QuerySearchResult` is a response class which has a hierarchy from `TransportResponse` which is passed as part of the message between the nodes. `Writeable` is used then in search service and actions to determine the kind of response. For example something like this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/SearchTransportService.java#L244.

Reply:
Thank you!

Reply:
@VachaShah  OK, I think we won't be able to move on with this change before we solve the protocol handling by outbound communication (we sort of addressed the inbound only, but not outbound, `TransportResponseHandler` & are heavily relying on "native" serialization).  In any case,  I think the serde should be moved away from the response classes, and handled by the `TransportResponseHandler`, depending on the protocol of the inbound message).

Reply:
I'm with this. I think the "right" way is to first PR splitting out the native serialization and deserialization out of `FetchSearchResult`. It's a hefty piece of work, but who dares wins.

Reply:
Makes sense! I will look into this. For abstracting the outbound side of transport and adding support for multiple protocols on that side of transport, I have created #13293. Next, I am looking into decoupling `Writeable` for native serialization from `TransportResponseHandler`.